### PR TITLE
VIM-769 Support for vim-surround

### PR DESCRIPTION
**Pull request by dhleong**
This is an initial stab at defining semantics for replicating vim plugin behavior within IdeaVim, with some preliminary support for tpope's excellent [vim-surround](https://github.com/tpope/vim-surround) plugin.

I'm not sure if there's a more idiomatic way of doing stuff like this, but since I didn't hear back on my request for guidance I just went ahead and made something up.

I created a `Plugin` interface which contains a `Registrar` in which plugins register. Each plugin has an "option name" which is used to enable the plugin using the vim options syntax, eg:

```vim
  set surround
```

Plugins can be activated at runtime using this syntax, but I wasn't sure how to go about disabling them (I think we'd need a hook to register and de-register actions, or perhaps make a subclass of VimCommandAction that won't be automatically registered at init). I'm not sure that's a huge concern, though, as generally you set up your plugins once in your vimrc and just go.

Since plugin bindings are optional and are enabled after init via .ideavimrc, I slightly refactored `ParentNode` to be able to correctly handle adding new `BranchNode` onto a `BranchNode` that already has an `ArgumentNode`. This way, plugins can override built-in bindings as expected, as well as expand on them.

The included vim-surround port supports most of the core surroundings, but currently leaves xml/html tag support to future work, as well as (I'm sure) a few more obscure features that I wasn't aware of, but I wanted to get this PR up and get some feedback on the approach before I try to figure that one out.

Since the `ys` and `cs` motions take more than just a motion, I mimicked vim-surround's implementation and added a `getChar()` command to `KeyHandler`. While necessarily asynchronous, taking a callback, this does block other vim input processing, as expected. If there's a better way to deal with this sort of thing I'd love to hear it, but this seemed like it would prove helpful in porting other vim plugins in the future.

UPDATE: Repeat support via `.` for `cs` and `ys` is achieved by subclassing `Argument` and storing our extra arguments in it (ie, the `SurroundPair` object). I think this is the simplest way to support the more complex surrounds, such as that achieved by `ysiwtdiv>` to surround the inner word with `<div></div>` tags (which is not yet implemented, but I'm thinking ahead for when we figure that out).

### CODE REVIEW

**Comment by expez** [comment]
Hot damn.  This is the plugins I miss the most in ideavim.  Thanks for taking this on!

**Comment by vlasovskikh** [comment]
@dhleong Thanks for your pull requests. I might be unavailable for a couple of weeks. I'll review your PRs later on.

**Comment by dhleong** [comment]
Cheers :) I'll probably keep working on them as time allows

**Comment by huyz** [comment]
awesome

**Comment by brcolow** [comment]
In reference to your comment about a more idiomatic way to do plugin registration...have you considered using a <a href="https://docs.oracle.com/javase/tutorial/ext/basics/spi.html#define-the-service-provider-interface">Service Provider Interface (SPI)</a>? Considering Plugin is an interface, it seems to me this would fit very well, or at least provide you with an alternative to a static array of Plugin objects - which to me (and as you clearly state) is less then ideal.

**Comment by dhleong** [comment]
I admit I'm unfamiliar with that API, but it seems like a good candidate if it plays well with IntelliJ. I'll wait to see if @vlasovskikh has any preferences before proceeding with anything though

**Comment by brcolow** [comment]
Certainly. I am very excited by your work here, and want to give you kudos =)

**Comment by dhleong** [comment]
Thanks :)

**Comment by tlnagy** [comment]
Any progress on this? This would be great feature to have. I use vim-surround a lot.

**Comment by dhleong** [comment]
@vlasovskikh Have you had a chance to take a look at this? Thoughts on how plugins should declare themselves?

**Comment by vlasovskikh** [comment]
@dhleong Not yet, sorry. I think there is no need in creating a Vim plugin mechanism for IdeaVim. Currently it looks like there might be only a few plugins worth "porting" to IdeaVim, so it would be simpler to just include them as a part of the core IdeaVim. If this situation changes, we will switch to some plugin system for that kind of features.

**Comment by dhleong** [comment]
Okay, sounds good, but I assume we want to be able to enable and disable
plugins in the .ideavimrc? I believe my initial implementation just uses a
map or an array or something, but @brcolow proposed using SPI, and I guess
IntelliJ has its own XML mechanism for providing feature implementations.
Since, as you said, there are only a few plugins that might get ported, an
array may be sufficient, but I'm open to something fancier. Just wanted
your input on a preferred method.

**Comment by portOdin** [comment]
What are the commands for using your surround implementation? I tried cs" and ys" but it ignores c and s just switches to input mode.

**Comment by dhleong** [comment]
`ds`, `cs`, and `ys` should all be supported as they are in [vim-surround](https://github.com/tpope/vim-surround) (insofar as the relevant motion is also supported in ideavim). See the [unit tests](https://github.com/dhleong/ideavim/blob/dhleong/Surround/test/org/jetbrains/plugins/ideavim/action/plugin/surround/SurroundPluginTest.java) for examples.

**Comment by portOdin** [comment]
Nevermind, I forgot to `set surround`. Great work, by the way.

**Comment by dhleong** [comment]
@vlasovskikh I think the biggest reason for requiring plugins be enable/disable-able is to be able to handle such plugins that override default bindings. vim-surround is innocuous enough, but something like [vim-sneak](https://github.com/justinmk/vim-sneak), for example, steals your `s`-key by default. I suppose for such plugins we could instead add support for `<Plug>FUNCTION` things and expose plugin functionality that way, but that might be a bit much to deal with. The `set` syntax, at least, mimics that of plugin managers like vim-plug. 

This doesn't address how plugins are declared in code, but perhaps we might prefer supporting the syntax from Vundle or vim-plug with the same github artifacts? That way any plugins a user might have in their .vimrc (if sourced as mentioned in the README) will be automatically enabled. Doesn't help pathogen users, of course, and there might be confusion if there's any difference between the vim plugin and the ideavim plugin (which I suspect there would be).... Just a thought.

**Comment by nsh87** [comment]
Merge please, this is a sought-after feature. It's listed as a ["Major" priority](https://youtrack.jetbrains.com/issue/VIM-769) and seems like @dhleong has done the work already.

**Comment by seanknox** [comment]
@dhleong @vlasovskikh What's needed to accept this PR? This is a widely used feature; [3000+ stars](https://github.com/tpope/vim-surround) on the original vim plugin repo alone.

**Comment by dhleong** [comment]
Waiting on @vlasovskikh to decide how he wants plugins to be provided/specified. However we do it here will inform how other such plugins are made available in IdeaVim....

**Comment by portOdin** [comment]
I switched systems, and can't remember how I set this up. Did you have an IdeaVim jar with this implemented, @dhleong ?

**Comment by dhleong** [comment]
I don't, but you should be able to make one by running `ant dist`

**Comment by shockone** [comment]
A workaround, add to `~/.ideavimrc`:
```
" Change surroundings.
nmap cs'" F'r"f'r"
nmap cs"' F"r'f"r'
nmap cs[( F[r(f]r)
```

**Comment by dhleong** [comment]
@vlasovskikh I've rebased on this off of master so it can clean merge; the tests all pass. Do you have any thoughts on this?

**Comment by tlnagy** [comment]
Great work @dhleong, I look forward to this feature.

**Comment by vlasovskikh** [comment]
@dhleong Going to review this PR very soon.

**Comment by vlasovskikh** [comment]
@dhleong I've done reviewing your PR and I've come up with a few ideas about the infrastructure for plugins for IdeaVim, see my PR #103. I believe with this framework in place it should be easier to port Vim plugins. My PR is a work in progress and I hope you'll join me. I would like to refactor your code so it fits into this plugin framework.

Here are key points of the review and ideas for the plugin infrastructure:

1. Lifecycle of plugins (better call them "IdeaVim extensions" since IdeaVim is a plugin itself)
  * There should be init() and dispose() methods (though it's not possible to disable the original Vim plugins as far as I know)
  * IntelliJ platform has a comprehensive extension mechanism, we can utilize it
2. Implementing IdeaVim extensions / porting Vim plugins
  * We should stay closer to the way the original plugins are written. In particular, the IdeaVim API for plugins should be as similar to the original Vim as possible
  * Registration of the extension functions should be done via `:map` commands, the original key handling loop should not be modified
  * There should be an API facade for the extension writers
3. Synchronous input functions
  * Async versions of `getchar()` necessary due to the async nature of the GUI events are very hard to follow and error-prone
  * It's possible to emulate sync input functions of Vim like `getchar()` and `input()` with modal Swing dialogs that look like the Vim command line
4. The repeat `.` command for the extension actions
  * Let's worry about the repeat command later, it doesn't work for many actions in the original Vim plugins

I'll describe the concrete steps I've taken in my PR #103.

**Comment by dhleong** [comment]
@vlasovskikh Thanks for looking into this in such depth! See my comments on your PR. I think we'll be closing this PR in favor of yours sometime soon. Would you also like my other PR #86 to be implemented as an extension?

**Comment by dhleong** [comment]
Going to go ahead and close this PR in favor of the work going on in PR #103
